File: C++/BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Graph {
public:
    int V;
    vector<vector<int>> adjList;
    vector<int> traversalPath;

    Graph(int V) {
        this->V = V;
        adjList.resize(V);
    }

    void addEdge(int src, int dest) {
        adjList[src - 1].push_back(dest);
        adjList[dest - 1].push_back(src); // For undirected graph, add the reverse edge
    }

    void BFS(int startVertex) {
        vector<bool> visited(V, false);
        queue<int> queue;

        visited[startVertex - 1] = true;
        queue.push(startVertex);

        while (!queue.empty()) {
            int currentVertex = queue.front();
            traversalPath.push_back(currentVertex);
            queue.pop();

            for (int adjVertex : adjList[currentVertex - 1]) {
                if (!visited[adjVertex - 1]) {
                    visited[adjVertex - 1] = true;
                    queue.push(adjVertex);
                }
            }
        }
    }

    void printGraph() {
        for (int v = 1; v <= V; v++) {
            cout << "Vertex " << v << ": ";
            for (int i = 1; i <= V; i++) {
                bool isNeighbor = false;
                for (int neighbor : adjList[v - 1]) {
                    if (neighbor == i || find(adjList[i - 1].begin(), adjList[i - 1].end(), v) != adjList[i - 1].end()) {
                        isNeighbor = true;
                        break;
                    }
                }
                cout << isNeighbor << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    int V, E, src, dest;

    cout << "Enter the number of vertices: ";
    cin >> V;
    Graph graph(V);

    cout << "Enter the number of edges: ";
    cin >> E;

    for (int i = 0; i < E; i++) {
        cout << "Enter edge (source destination): ";
        cin >> src >> dest;
        graph.addEdge(src, dest);
    }

    cout << "Starting vertex for BFS: ";
    int startVertex;
    cin >> startVertex;

    graph.BFS(startVertex);

    cout << "Breadth-First Search (BFS) starting from vertex " << startVertex << ":\n";
    cout << "Traversal path: ";
    for (int i = 0; i < graph.traversalPath.size(); i++) {
        cout << graph.traversalPath[i];
        if (i < graph.traversalPath.size() - 1) {
            cout << " -> ";
        }
    }
    cout << endl;

    cout << "Adjacency Graph:\n";
    graph.printGraph();

    return 0;
}

File: Python/BFS.py
from collections import deque

class Graph:
    def __init__(self, V):
        self.V = V
        self.adjList = [[] for _ in range(V)]
        self.traversalPath = []

    def addEdge(self, src, dest):
        self.adjList[src - 1].append(dest)
        self.adjList[dest - 1].append(src)  # For undirected graph, add the reverse edge

    def BFS(self, startVertex):
        visited = [False] * self.V
        queue = deque()

        visited[startVertex - 1] = True
        queue.append(startVertex)

        while queue:
            currentVertex = queue.popleft()
            self.traversalPath.append(currentVertex)

            for adjVertex in self.adjList[currentVertex - 1]:
                if not visited[adjVertex - 1]:
                    visited[adjVertex - 1] = True
                    queue.append(adjVertex)

    def printGraph(self):
        for v in range(1, self.V + 1):
            print(f"Vertex {v}: ", end='')
            for i in range(1, self.V + 1):
                isNeighbor = False
                for neighbor in self.adjList[v - 1]:
                    if neighbor == i or v in self.adjList[i - 1]:
                        isNeighbor = True
                        break
                print(int(isNeighbor), end=' ')
            print()

if __name__ == "__main__":
    V = int(input("Enter the number of vertices: "))
    graph = Graph(V)

    E = int(input("Enter the number of edges: "))

    for _ in range(E):
        src, dest = map(int, input("Enter edge (source destination): ").split())
        graph.addEdge(src, dest)

    startVertex = int(input("Starting vertex for BFS: "))
    graph.BFS(startVertex)

    print(f"Breadth-First Search (BFS) starting from vertex {startVertex}:")
    print("Traversal path:", " -> ".join(map(str, graph.traversalPath)))

    print("Adjacency Graph:")
    graph.printGraph()
